File: web/src/ir.ts
// Pattern IR — TypeScript side of the wire format decoded by `crates/muse-core/src/ir.rs`.
//
// Keep the encoder below in sync with the Rust decoder:
//
//   magic "MUSE" | u32 version=1 | node
//   node: u8 tag + payload (see ir.rs header for the tag table)

export const NCTL = 12;

/** Control slot indices — must match `ir.rs`. */
export const CTL = {
  note: 0,
  gain: 1,
  cutoff: 2,
  pan: 3,
  attack: 4,
  decay: 5,
  sustain: 6,
  release: 7,
  delay: 8,
  resonance: 9,
  speed: 10,
  crush: 11,
} as const;

export type Ctl = (number | null)[]; // length NCTL, null = unset (NaN on the wire)

export type Pat =
  | { t: "rest" }
  | { t: "atom"; sound: string; ctl: Ctl }
  | { t: "cat"; kids: Pat[] }
  | { t: "overlay"; kids: Pat[] }
  | { t: "fast"; k: number; kid: Pat }
  | { t: "every"; n: number; step: Pat; kid: Pat }
  | { t: "sometimes"; p: number; step: Pat; kid: Pat }
  | { t: "rev"; kid: Pat }
  | { t: "altern"; kids: Pat[] }
  | { t: "struct"; mask: string; kid: Pat }
  | { t: "shift"; d: number; kid: Pat }
  | { t: "setctl"; slot: number; v: number; kid: Pat }
  | { t: "setsound"; sound: string; kid: Pat }
  | { t: "addctl"; slot: number; d: number; kid: Pat }
  | { t: "chunk"; n: number; step: Pat; kid: Pat };

export const rest: Pat = { t: "rest" };

export function emptyCtl(): Ctl {
  return new Array<number | null>(NCTL).fill(null);
}

export function atom(sound: string, note?: number): Pat {
  const ctl = emptyCtl();
  if (note !== undefined) ctl[CTL.note] = note;
  return { t: "atom", sound, ctl };
}

class Writer {
  private buf = new Uint8Array(4096);
  private len = 0;

  private ensure(n: number) {
    if (this.len + n <= this.buf.length) return;
    let cap = this.buf.length * 2;
    while (cap < this.len + n) cap *= 2;
    const next = new Uint8Array(cap);
    next.set(this.buf.subarray(0, this.len));
    this.buf = next;
  }

  u8(v: number) {
    this.ensure(1);
    this.buf[this.len++] = v & 0xff;
  }

  u32(v: number) {
    this.ensure(4);
    new DataView(this.buf.buffer).setUint32(this.len, v >>> 0, true);
    this.len += 4;
  }

  u64(v: number) {
    this.ensure(8);
    new DataView(this.buf.buffer).setBigUint64(this.len, BigInt(Math.trunc(v)), true);
    this.len += 8;
  }

  f64(v: number) {
    this.ensure(8);
    new DataView(this.buf.buffer).setFloat64(this.len, v, true);
    this.len += 8;
  }

  str(s: string) {
    const bytes = new TextEncoder().encode(s);
    this.u32(bytes.length);
    this.ensure(bytes.length);
    this.buf.set(bytes, this.len);
    this.len += bytes.length;
  }

  finish(): Uint8Array {
    return this.buf.slice(0, this.len);
  }
}

const TAG = {
  rest: 0,
  atom: 1,
  cat: 2,
  overlay: 3,
  fast: 4,
  every: 6,
  rev: 7,
  altern: 8,
  struct: 9,
  setctl: 11,
  setsound: 12,
  shift: 13,
  sometimes: 14,
  addctl: 15,
  chunk: 16,
} as const;

function node(w: Writer, p: Pat): void {
  switch (p.t) {
    case "rest":
      w.u8(TAG.rest);
      return;
    case "atom": {
      w.u8(TAG.atom);
      w.str(p.sound);
      for (let i = 0; i < NCTL; i++) {
        const v = p.ctl[i];
        w.f64(v === null || v === undefined || Number.isNaN(v) ? NaN : v);
      }
      return;
    }
    case "cat":
      w.u8(TAG.cat);
      writeKids(w, p.kids);
      return;
    case "overlay":
      w.u8(TAG.overlay);
      writeKids(w, p.kids);
      return;
    case "fast":
      w.u8(TAG.fast);
      w.f64(p.k);
      node(w, p.kid);
      return;
    case "every":
      w.u8(TAG.every);
      w.u64(p.n);
      node(w, p.step);
      node(w, p.kid);
      return;
    case "sometimes":
      w.u8(TAG.sometimes);
      w.f64(p.p);
      node(w, p.step);
      node(w, p.kid);
      return;
    case "rev":
      w.u8(TAG.rev);
      node(w, p.kid);
      return;
    case "altern":
      w.u8(TAG.altern);
      writeKids(w, p.kids);
      return;
    case "struct":
      w.u8(TAG.struct);
      w.str(p.mask);
      node(w, p.kid);
      return;
    case "shift":
      w.u8(TAG.shift);
      w.f64(p.d);
      node(w, p.kid);
      return;
    case "setctl":
      w.u8(TAG.setctl);
      w.u8(p.slot);
      w.f64(p.v);
      node(w, p.kid);
      return;
    case "setsound":
      w.u8(TAG.setsound);
      w.str(p.sound);
      node(w, p.kid);
      return;
    case "addctl":
      w.u8(TAG.addctl);
      w.u8(p.slot);
      w.f64(p.d);
      node(w, p.kid);
      return;
    case "chunk":
      w.u8(TAG.chunk);
      w.u64(p.n);
      node(w, p.step);
      node(w, p.kid);
      return;
    default: {
      const never: never = p;
      throw new Error(`unknown pattern node: ${JSON.stringify(never)}`);
    }
  }
}

function writeKids(w: Writer, kids: Pat[]) {
  if (kids.length === 0 || kids.length > 1024) {
    throw new Error(`bad child count ${kids.length}`);
  }
  w.u32(kids.length);
  for (const k of kids) node(w, k);
}

/** Encode a pattern to the wasm-side IR buffer. */
export function encode(pat: Pat): Uint8Array {
  const w = new Writer();
  w.u8(0x4d);
  w.u8(0x55);
  w.u8(0x53);
  w.u8(0x45);
  w.u32(1);
  node(w, pat);
  return w.finish();
}

// ---------------------------------------------------------------------------
// Packed scheduler records (sched.rs `pack_range`)
// ---------------------------------------------------------------------------

export const EVENT_HEADER = 8 + 8 + 8 * NCTL; // onset f64, dur f64, ctl[12] f64

export interface SchedEvent {
  onsetCycle: number;
  durSec: number;
  ctl: Float64Array; // defaults already applied
  sound: string;
}

/** Parse packed event records written by `sched_query` / `sched_peek`. */
export function unpackEvents(buf: ArrayBuffer | Uint8Array): SchedEvent[] {
  const bytes = buf instanceof Uint8Array ? buf : new Uint8Array(buf);
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const out: SchedEvent[] = [];
  let i = 0;
  while (i + EVENT_HEADER + 2 <= bytes.length) {
    const onsetCycle = dv.getFloat64(i, true);
    const durSec = dv.getFloat64(i + 8, true);
    i += 16;
    const ctl = new Float64Array(NCTL);
    for (let s = 0; s < NCTL; s++) {
      ctl[s] = dv.getFloat64(i, true);
      i += 8;
    }
    const sl = dv.getUint16(i, true);
    i += 2;
    if (i + sl > bytes.length) break;
    const sound = new TextDecoder().decode(bytes.subarray(i, i + sl));
    i += sl;
    out.push({ onsetCycle, durSec, ctl, sound });
  }
  return out;
}

File: crates/muse-core/src/ir.rs
//! Pattern IR: the data structure that crosses the WASM boundary, plus its
//! binary decoder. The encoder lives in `web/src/ir.ts` — the two must agree
//! on the wire format:
//!
//! ```text
//! magic  "MUSE"            4 bytes
//! version u32 LE           1
//! node:
//!   tag  u8
//!   0 Rest                                   —
//!   1 Atom   u32 len, utf8, f64[12] ctl      (NaN = unset)
//!   2 Cat    u32 n, node*n                   sequence (mini-notation `a b`, `[..]`)
//!   3 Overlay u32 n, node*n                  stack    (mini-notation `(..)`)
//!   4 Fast   f64 k, node                     k>1 faster, k<1 slower
//!   5 Slow   f64 k, node                     sugar, decoded as Fast(1/k)
//!   6 Every  i64 n, node step, node child
//!   7 Rev    node
//!   8 Altern u32 n, node*n                   mini-notation `<a b>`
//!   9 Struct u32 mlen, mask utf8, node child ('x' hit / '.' rest)
//!  10 Chunk  i64 n, i64 idx, node child
//!  11 SetCtl u8 slot, f64 v, node child
//!  12 SetSound u32 len, utf8, node child
//!  13 Shift  f64 cycles, node child
//!  14 Sometimes f64 prob, node step, node child
//!  15 AddCtl  u8 slot, f64 delta, node child    (transpose-style, NaN -> default)
//!  16 Chunk   u64 n, node step, node child      (cycle split into n slots,
//!                                                 slot `cycle % n` gets `step`)
//! ```

/// Number of control slots carried by every event.
pub const NCTL: usize = 12;

pub const CTL_NOTE: usize = 0;
pub const CTL_GAIN: usize = 1;
pub const CTL_CUTOFF: usize = 2;
pub const CTL_PAN: usize = 3;
pub const CTL_ATTACK: usize = 4;
pub const CTL_DECAY: usize = 5;
pub const CTL_SUSTAIN: usize = 6;
pub const CTL_RELEASE: usize = 7;
pub const CTL_DELAY: usize = 8;
pub const CTL_RESONANCE: usize = 9;
pub const CTL_SPEED: usize = 10;
pub const CTL_CRUSH: usize = 11;

/// Defaults applied to `NaN` slots when an event is packed for playback.
pub const DEFAULTS: [f64; NCTL] = [
    60.0,    // note (midi)
    1.0,     // gain
    12000.0, // cutoff Hz
    0.5,     // pan
    0.005,   // attack
    0.05,    // decay
    1.0,     // sustain
    0.05,    // release
    0.0,     // delay send
    0.1,     // resonance
    1.0,     // speed
    0.0,     // crush
];

#[derive(Clone, Debug)]
pub enum Pat {
    /// No events.
    Rest,
    /// One event per cycle, occupying the whole cycle in its own frame.
    Atom {
        sound: String,
        ctl: [f64; NCTL],
    },
    /// Divide the cycle evenly among children (`a b`, `[a b]`).
    Cat(Vec<Pat>),
    /// All children in parallel (`(a b)`).
    Overlay(Vec<Pat>),
    /// Time scale: output time = input time / k.
    Fast(f64, Box<Pat>),
    /// Apply `step` on cycles where `cycle % n == 0`, base pattern otherwise.
    Every(i64, Box<Pat>, Box<Pat>),
    /// Apply `step` on cycles passing a deterministic probability test.
    Sometimes(f64, Box<Pat>, Box<Pat>),
    /// Mirror each cycle.
    Rev(Box<Pat>),
    /// One child per cycle (`<a b>`).
    Altern(Vec<Pat>),
    /// Intersect the child with a hit/rest mask (euclid, struct, chunk..).
    Struct {
        mask: Vec<bool>,
        child: Box<Pat>,
    },
    /// Offset in cycles.
    Shift(f64, Box<Pat>),
    /// Force a control slot on all descendant events (wins over atom values).
    SetCtl(usize, f64, Box<Pat>),
    /// Force the sound on all descendant events.
    SetSound(String, Box<Pat>),
    /// Add `delta` to a control slot (NaN treated as the default first).
    AddCtl(usize, f64, Box<Pat>),
    /// Divide each cycle into `n` slots; apply `step` to slot `cycle % n`.
    Chunk(i64, Box<Pat>, Box<Pat>),
}

struct Reader<'a> {
    b: &'a [u8],
    i: usize,
}

impl<'a> Reader<'a> {
    fn take(&mut self, n: usize) -> Result<&'a [u8], String> {
        if self.i + n > self.b.len() {
            return Err(format!("unexpected end of IR at byte {}", self.i));
        }
        let s = &self.b[self.i..self.i + n];
        self.i += n;
        Ok(s)
    }

    fn u8(&mut self) -> Result<u8, String> {
        Ok(self.take(1)?[0])
    }

    fn u32(&mut self) -> Result<u32, String> {
        Ok(u32::from_le_bytes(self.take(4)?.try_into().unwrap()))
    }

    fn u64(&mut self) -> Result<u64, String> {
        Ok(u64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }

    fn f64(&mut self) -> Result<f64, String> {
        Ok(f64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }

    fn str(&mut self) -> Result<String, String> {
        let n = self.u32()? as usize;
        if n > 1 << 20 {
            return Err("string too long".into());
        }
        String::from_utf8(self.take(n)?.to_vec()).map_err(|_| "invalid utf8".into())
    }
}

fn node(r: &mut Reader, depth: usize) -> Result<Pat, String> {
    if depth > 64 {
        return Err("pattern nesting too deep".into());
    }
    let tag = r.u8()?;
    match tag {
        0 => Ok(Pat::Rest),
        1 => {
            let sound = r.str()?;
            let mut ctl = [f64::NAN; NCTL];
            for c in ctl.iter_mut() {
                *c = r.f64()?;
            }
            Ok(Pat::Atom { sound, ctl })
        }
        2 => Ok(Pat::Cat(children(r, depth)?)),
        3 => Ok(Pat::Overlay(children(r, depth)?)),
        4 => {
            let k = r.f64()?;
            Ok(Pat::Fast(k, Box::new(node(r, depth + 1)?)))
        }
        5 => {
            let k = r.f64()?;
            let inner = node(r, depth + 1)?;
            Ok(Pat::Fast(if k.abs() < 1e-9 { 1e9 } else { 1.0 / k }, Box::new(inner)))
        }
        6 => {
            let n = r.u64()? as i64;
            let step = Box::new(node(r, depth + 1)?);
            let child = Box::new(node(r, depth + 1)?);
            Ok(Pat::Every(n.max(1), step, child))
        }
        7 => Ok(Pat::Rev(Box::new(node(r, depth + 1)?))),
        8 => Ok(Pat::Altern(children(r, depth)?)),
        9 => {
            let mask_str = r.str()?;
            if mask_str.is_empty() || mask_str.len() > 4096 {
                return Err("bad struct mask".into());
            }
            let mask = mask_str.bytes().map(|c| c != b'.').collect();
            Ok(Pat::Struct { mask, child: Box::new(node(r, depth + 1)?) })
        }
        10 => {
            // Chunk(n, idx) decoded as a single-slot struct: cheaper and
            // shares the intersection logic.
            let n = r.u64()? as i64;
            if n <= 0 || n > 4096 {
                return Err("bad chunk count".into());
            }
            let idx = (r.u64()? as i64).rem_euclid(n);
            let child = Box::new(node(r, depth + 1)?);
            let mask = (0..n).map(|i| i == idx).collect();
            Ok(Pat::Struct { mask, child })
        }
        11 => {
            let slot = r.u8()? as usize;
            let v = r.f64()?;
            if slot >= NCTL {
                return Err(format!("ctl slot {slot} out of range"));
            }
            Ok(Pat::SetCtl(slot, v, Box::new(node(r, depth + 1)?)))
        }
        12 => {
            let s = r.str()?;
            Ok(Pat::SetSound(s, Box::new(node(r, depth + 1)?)))
        }
        13 => {
            let d = r.f64()?;
            Ok(Pat::Shift(d, Box::new(node(r, depth + 1)?)))
        }
        14 => {
            let p = r.f64()?;
            let step = Box::new(node(r, depth + 1)?);
            let child = Box::new(node(r, depth + 1)?);
            Ok(Pat::Sometimes(p.clamp(0.0, 1.0), step, child))
        }
        15 => {
            let slot = r.u8()? as usize;
            let delta = r.f64()?;
            if slot >= NCTL {
                return Err(format!("ctl slot {slot} out of range"));
            }
            Ok(Pat::AddCtl(slot, delta, Box::new(node(r, depth + 1)?)))
        }
        16 => {
            let n = r.u64()? as i64;
            if n <= 0 || n > 4096 {
                return Err("bad chunk count".into());
            }
            let step = Box::new(node(r, depth + 1)?);
            let child = Box::new(node(r, depth + 1)?);
            Ok(Pat::Chunk(n, step, child))
        }
        t => Err(format!("unknown IR tag {t}")),
    }
}

fn children(r: &mut Reader, depth: usize) -> Result<Vec<Pat>, String> {
    let n = r.u32()? as usize;
    if n == 0 || n > 1024 {
        return Err(format!("bad child count {n}"));
    }
    (0..n).map(|_| node(r, depth + 1)).collect()
}

pub fn decode(bytes: &[u8]) -> Result<Pat, String> {
    let mut r = Reader { b: bytes, i: 0 };
    if r.take(4)? != b"MUSE" {
        return Err("bad magic".into());
    }
    let v = r.u32()?;
    if v != 1 {
        return Err(format!("unsupported IR version {v}"));
    }
    let p = node(&mut r, 0)?;
    Ok(p)
}
